sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotMarkEnumsWithFlags.cs

```diff
@@ -0,0 +1,130 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotMarkEnumsWithFlags : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4070";
+        private const string MessageFormat = "Remove the 'FlagsAttribute' from this enum.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var enumDeclaration = (EnumDeclarationSyntax)c.Node;
+                    var enumSymbol = c.SemanticModel.GetDeclaredSymbol(enumDeclaration);
+
+                    if (!enumDeclaration.HasFlagsAttribute(c.SemanticModel) ||
+                        enumDeclaration.Identifier.IsMissing ||
+                        enumSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var membersWithValues = enumSymbol.GetMembers()
+                        .OfType<IFieldSymbol>()
+                        .Select(member => new { Member = member, Value = GetEnumValueOrDefault(member) })
+                        .OrderByDescending(tuple => tuple.Value)
```

Comment:
If you mean elements with `Value = null` then no we need to keep them because we need to report on them.